File: examples/analyse_pgn.py
from __future__ import annotations

from pathlib import Path

import spooky_chess

PGN_PATH = Path("pgn/example/scholars_mate.pgn")
DEPTH = 10
ENGINE_PATH = "stockfish"


def print_analysis(position_index: int, result: spooky_chess.SearchResult) -> None:
    print(
        f"  position {position_index:>3}: "
        f"bestmove={result.best_move_lan} "
        f"depth={result.depth} "
        f"score_cp={result.score_cp} "
        f"score_mate={result.score_mate} "
        f"pv={' '.join(result.pv)}"
    )


def main() -> None:
    games = spooky_chess.parse_pgn(PGN_PATH.read_text())
    print(f"Loaded {len(games)} game(s) from {PGN_PATH}")

    engine = spooky_chess.UciEngine(ENGINE_PATH)
    engine.set_option("Hash", "32")
    engine.is_ready()

    try:
        for game_index, pgn_game in enumerate(games, start=1):
            print()
            print(f"Game {game_index}: {pgn_game.white() or '?'} vs {pgn_game.black() or '?'} ({pgn_game.result()})")

            engine.set_position_pgn_start(pgn_game)

            for ply_index, played_move in enumerate(pgn_game.moves()):
                # Get best move
                result = engine.go_depth(DEPTH)
                print_analysis(ply_index, result)

                # Play played move
                engine_ok = engine.make_move(played_move)
                if not engine_ok:
                    raise RuntimeError(f"failed to replay PGN move {played_move.to_lan()} at ply {ply_index + 1}")

            if engine.is_over():
                print(f"  position {len(pgn_game.moves()):>3}: terminal position, no legal best move")
            else:
                final_result = engine.go_depth(DEPTH)
                print_analysis(len(pgn_game.moves()), final_result)
    finally:
        engine.quit()


if __name__ == "__main__":
    main()

File: examples/analyse_pgn.rs
use std::error::Error;
use std::fs;

use spooky_chess::pgn::parse_pgn;
use spooky_chess::uci::{SearchResult, UciEngine};

const PGN_PATH: &str = "pgn/example/scholars_mate.pgn";
const DEPTH: u32 = 10;
const ENGINE_PATH: &str = "stockfish";

fn print_analysis(position_index: usize, result: &SearchResult) {
    let info = result.info.last();
    let depth = info.and_then(|line| line.depth);
    let score_cp = info.and_then(|line| line.score_cp);
    let score_mate = info.and_then(|line| line.score_mate);
    let pv = info.map(|line| line.pv.join(" ")).unwrap_or_default();

    println!(
        "  position {:>3}: bestmove={} depth={depth:?} score_cp={score_cp:?} score_mate={score_mate:?} pv={pv}",
        position_index, result.best_move_lan
    );
}

fn main() -> Result<(), Box<dyn Error>> {
    let pgn = fs::read_to_string(PGN_PATH)?;
    let games = parse_pgn(&pgn)?;
    println!("Loaded {} game(s) from {PGN_PATH}", games.len());

    let mut engine = UciEngine::new(ENGINE_PATH, &[])?;
    engine.set_option("Hash", "32")?;
    engine.is_ready()?;

    for (game_index, pgn_game) in games.iter().enumerate() {
        println!();
        println!(
            "Game {}: {} vs {} ({})",
            game_index + 1,
            pgn_game.headers.white().unwrap_or("?"),
            pgn_game.headers.black().unwrap_or("?"),
            pgn_game.result
        );

        engine.set_position_pgn_start(pgn_game)?;

        for (ply_index, played_move) in pgn_game.moves.iter().enumerate() {
            // Get best move
            let search = engine.go_depth(DEPTH)?;
            print_analysis(ply_index, &search);

            // Play played move
            let engine_ok = engine.make_move(played_move)?;
            if !engine_ok {
                return Err(format!(
                    "failed to replay PGN move {} at ply {}",
                    played_move.to_lan(),
                    ply_index + 1
                )
                .into());
            }
        }

        if engine.is_over() {
            println!(
                "  position {:>3}: terminal position, no legal best move",
                pgn_game.moves.len()
            );
        } else {
            let final_search = engine.go_depth(DEPTH)?;
            print_analysis(pgn_game.moves.len(), &final_search);
        }
    }

    engine.quit()?;
    Ok(())
}
